src/SapNwRfc/Internal/Dynamic/DynamicRfcFunction.cs

```diff
@@ -0,0 +1,158 @@
+using System.Collections.Generic;
+using System.Dynamic;
+using SapNwRfc.Internal.Interop;
+
+namespace SapNwRfc.Internal.Dynamic
+{
+    internal sealed class DynamicRfcFunction : DynamicObject, IReadOnlyDictionary<string, object>
+    {
+        private readonly RfcInterop _interop;
+        private readonly IntPtr _functionHandle;
+        private readonly ISapFunctionMetadata _functionMetadata;
+
+        internal DynamicRfcFunction(RfcInterop interop, IntPtr functionHandle, ISapFunctionMetadata functionMetadata)
+        {
+            _interop = interop;
+            _functionHandle = functionHandle;
+            _functionMetadata = functionMetadata;
+        }
+
+        private bool TryGetByName(string name, Type returnType, out object result)
+        {
+            if (_functionMetadata.Parameters.TryGetValue(name, out ISapParameterMetadata parameter))
+            {
+                if (DynamicRfc.TryGetRfcValue(_interop, _functionHandle, parameter.Name, parameter.Type, parameter.GetTypeMetadata, returnType, out result))
+                {
+                    return true;
+                }
+            }
+
+            result = null;
+            return false;
+        }
+
+        private bool TryGetByIndex(int index, Type returnType, out object result)
+        {
+            if (index >= 0 && index < _functionMetadata.Parameters.Count)
+            {
+                ISapParameterMetadata parameter = _functionMetadata.Parameters[index];
+
+                if (DynamicRfc.TryGetRfcValue(_interop, _functionHandle, parameter.Name, parameter.Type, parameter.GetTypeMetadata, returnType, out result))
+                {
+                    return true;
+                }
+            }
+
+            result = null;
+            return false;
+        }
+
+        public override IEnumerable<string> GetDynamicMemberNames()
+        {
+            return Keys;
+        }
+
+        public override bool TryGetMember(GetMemberBinder binder, out object result)
+        {
+            return TryGetByName(binder.Name, binder.ReturnType, out result);
+        }
+
+        public override bool TryGetIndex(GetIndexBinder binder, object[] indexes, out object result)
+        {
+            if (indexes.Length != 1)
+                throw new ArgumentOutOfRangeException(nameof(indexes));
+
+            var index = indexes[0];
+
+            if (index is string name)
+                return TryGetByName(name, binder.ReturnType, out result);
+
+            if (index is int i)
+                return TryGetByIndex(i, binder.ReturnType, out result);
+
+            if (index is uint u)
+                return TryGetByIndex((int)u, binder.ReturnType, out result);
+
+            throw new ArgumentException(nameof(indexes));
+        }
+
+        public override bool TryInvokeMember(InvokeMemberBinder binder, object[] args, out object result)
+        {
+            if (binder.Name == "GetMetadata" && binder.CallInfo.ArgumentCount == 0)
```

Comment:
With this code `GetMetadata` can be called on the results to get the current metadata. Not sure if we should expose it like this or just hide it.

```c#
var result = fn.Invoke<dynamic>();

result.GetMetadata();
result.MY_STRUCT.GetMetadata();
result.MY_TABLE.GetMetadata();
result.MY_TABLE[0].GetMetadata();
```